tests/topotests/lib/common_config.py

```diff
@@ -469,138 +498,150 @@ def reset_config_on_routers(tgen, routerName=None):
-
-        f = open(dname, "r")
-        delta = StringIO()
-        delta.write("configure terminal\n")
-        t_delta = f.read()
-
-        # Don't disable debugs
-        check_debug = True
-
-        for line in t_delta.split("\n"):
-            line = line.strip()
-            if line == "Lines To Delete" or line == "===============" or not line:
-                continue
-
-            if line == "Lines To Add":
-                check_debug = False
-                continue
-
-            if line == "============" or not line:
-                continue
-
-            # Leave debugs and log output alone
-            if check_debug:
-                if "debug" in line or "log file" in line:
-                    continue
-
-            delta.write(line)
-            delta.write("\n")
-
-        f.close()
-
-        delta.write("end\n")
-
-        output = router.vtysh_multicmd(delta.getvalue(), pretty_output=False)
-
-        delta.close()
-        delta = StringIO()
-        cfg = router.run("vtysh -c 'show running'")
-        for line in cfg.split("\n"):
-            line = line.strip()
-            delta.write(line)
-            delta.write("\n")
-
-        # Router current configuration to log file or console if
-        # "show_router_config" is defined in "pytest.ini"
-        if show_router_config:
-            logger.info("Configuration on router {} after reset:".format(rname))
-            logger.info(delta.getvalue())
-        delta.close()
+    for rname, p in procs.items():
+        _, error = p.communicate()
+        if p.returncode:
+            logger.error("Get running config for %s failed %d: %s", rname, p.returncode, error)
+            raise InvalidCLIError("vtysh show running error on {}: {}".format(rname, error))
+
+    #
+    # Get all delta's in parallel
+    #
+    procs = {}
+    for rname in router_list:
+        logger.info("Generating delta for router %s to new configuration", rname)
+        procs[rname] = Popen(
+            [ "/usr/lib/frr/frr-reload.py",
+              "--test-reset",
+              "--input",
+              run_cfg_fmt.format(rname),
+              "--test",
+              init_cfg_fmt.format(rname) ],
+            stdin=None,
+            stdout=open(delta_fmt.format(rname), "w"),
+            stderr=SUB_PIPE,
+        )
+    for rname, p in procs.items():
+        _, error = p.communicate()
+        if p.returncode:
+            logger.error("Delta file creation for %s failed %d: %s", rname, p.returncode, error)
+            raise InvalidCLIError("frr-reload error for {}: {}".format(rname, error))
+        # else:
+        #     logger.info("Delta file creation for %s: info: %s", rname, error)
+
+
+    #
+    # Apply all the deltas in parallel
+    #
+    procs = {}
+    for rname in router_list:
+        logger.info("Applying delta config on router %s", rname)
+
+        procs[rname] = router_list[rname].popen(
+            ["/usr/bin/env", "vtysh", "-f", delta_fmt.format(rname)],
```

Comment:
Will this cause any race condition issues as we are applying delta configs parallelly? Especially in large configurations.

Reply:
Not unless the API is mis-used. If the test needs things applied in a certain order then it needs to do that explicitly, not counting on a side-effect of the internals of the API (or what they were the day the person wrote the test).